ngraph/core/src/op/gather_elements.cpp

```diff
@@ -0,0 +1,133 @@
+{
+    // check types of input tensors
+    const auto& data_type = get_input_element_type(0);
+    const auto& indices_type = get_input_element_type(1);
+
+    NODE_VALIDATION_CHECK(this,
+                          indices_type == element::Type_t::i32 ||
+                              indices_type == element::Type_t::i64,
+                          "indices mush be of int32 or int64 type. But instead got: ",
+                          indices_type);
+
+    // check ranks of input tensors
+    const auto& data_pshape = get_input_partial_shape(0);
+    const auto& indices_pshape = get_input_partial_shape(1);
+    auto data_rank = data_pshape.rank();
+    auto indices_rank = indices_pshape.rank();
+
+    if (data_rank.is_static())
+    {
+        int64_t data_rank_size = data_rank.get_length();
+
+        NODE_VALIDATION_CHECK(this, data_rank_size > 1, "data rank must be greater than 1.");
+
+        if (m_axis < 0)
+        {
+            NODE_VALIDATION_CHECK(
+                this,
+                (-data_rank_size < m_axis) && (m_axis < data_rank_size - 1),
+                "axis must be within interval (-data.rank,  data.rank - 1). But instead Got: ",
+                m_axis);
+            m_axis = data_rank_size + m_axis;
+        }
+    }
+
+    if (indices_rank.is_static())
+    {
+        NODE_VALIDATION_CHECK(
+            this, indices_rank.get_length() > 1, "indices rank must be greater than 1.");
+    }
+
+    if (indices_rank.is_static() && data_rank.is_static())
+    {
+        NODE_VALIDATION_CHECK(this,
+                              data_rank.get_length() == indices_rank.get_length(),
+                              "data and indices rank must be equal. But instead got: ",
+                              data_rank.get_length(),
+                              " and ",
+                              indices_rank.get_length());
+
+        PartialShape out_shape_info(indices_pshape);
+
+        for (int i = 0; i < indices_rank.get_length(); i++)
+        {
+            if (i != m_axis && data_pshape[i].is_static() && indices_pshape[i].is_static())
+            {
+                NODE_VALIDATION_CHECK(
+                    this,
+                    data_pshape[i] == indices_pshape[i],
+                    "Sizes ",
+                    data_pshape[i],
+                    " and ",
+                    indices_pshape[i],
+                    " on axis ",
+                    i,
+                    " do not match. data and indices must have equal shapes except for axis ",
+                    m_axis);
+            }
+            // if indices.size of the current axis is unknown try to retrieve it from data
+            // if data_shape = {4, 4, ?} indices_shape = {1, ?, 5} and axis = 0
+            // make optimistic assumptions that data size along 2nd dim is also 5
+            else if (indices_pshape[i].is_dynamic() && data_pshape[i].is_static())
+            {
+                out_shape_info[i] = data_pshape[i];
+            }
```

Comment:
I see one more else:
Probably, in case of dynamic dimensions we have defined range (lower_bound, upper_bound) from both inputs and we need to check that these intervals intercepted and select one with minimal length.
The range should be also propagated into the output.

Reply:
validate_and_infer has been rewritten. Now intervals are taken into account. Please review